src/relocate.rs

```diff
@@ -71,67 +71,69 @@ pub(super) unsafe fn relocate(envp: *mut *mut u8) {
         current_aux = current_aux.add(1);
 
         match a_type as _ {
+            AT_BASE => auxv_base = a_val.addr(),
             AT_PAGESZ => auxv_page_size = a_val.addr(),
-            AT_PHDR => auxv_phdr = a_val.cast::<Elf_Phdr>(),
-            AT_PHNUM => auxv_phnum = a_val.addr(),
-            AT_PHENT => auxv_phent = a_val.addr(),
             AT_ENTRY => auxv_entry = a_val.addr(),
             AT_NULL => break,
             _ => (),
         }
     }
 
+    // There are four cases to consider here:
+    //
+    // 1. Static executable
+    //    This is the trivial case. No relocations necessary.
+    // 2. Static pie executable
+    //    We need to relocate ourself. `AT_PHDR` points to our own program
+    //    headers and `AT_ENTRY` to our own entry point.
+    // 3. Dynamic pie executable with external dynamic linker
+    //    We don't need to relocate ourself as the dynamic linker has already
+    //    done this. `AT_PHDR` points to our own program headers and `AT_ENTRY`
+    //    to our own entry point. `AT_BASE` contains the relocation offset of
+    //    the dynamic linker.
+    // 4. Shared library acting as dynamic linker.
+    //    We do need to relocate ourself. `AT_PHDR` doesn't point to our own
+    //    program headers and `AT_ENTRY` doesn't point to our own entry point.
+    //    `AT_BASE` contains our own relocation offset.
+
     // Compute the static address of `_start`.
     let static_start = compute_static_start();
 
-    // Our offset is the difference between the static address of `_start` and
-    // the actual runtime address of `_start` as given to us as the entry
-    // address in the AUX table.
-    let offset = auxv_entry.wrapping_sub(static_start);
+    if auxv_base != 0 && static_start == auxv_entry {
```

Comment:
I think the `auxv_base != 0` part of this can be removed. If `auxv_base` is 0, then the `if` at line 111 is true, `offset` is `auxv_entry.wrapping_sub(static_start)`, and the `offset == 0` check at line 125 is equivalent to testing `static_start == auxv_entry`.